messaging/src/test/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventIntegrationTest.java

```diff
@@ -456,16 +468,17 @@ void publishEventsAndProcessDeadLettersConcurrentlyInBulkShouldWorkFine() throws
         });
 
         startProcessingEvent();
-        processAnyDeadLettersPeriodically();
-
-        publishingThread.start();
 
         assertWithin(1, TimeUnit.SECONDS, () -> assertEquals(1, streamingProcessor.processingStatus().size()));
-        assertWithin(2, TimeUnit.SECONDS, () -> {
-            OptionalLong optionalPosition = streamingProcessor.processingStatus().get(0).getCurrentPosition();
+        publishingThread.start();
+        assertWithin(10, TimeUnit.SECONDS, () -> {
+            EventTrackerStatus eventTrackerStatus = streamingProcessor.processingStatus().get(0);
+            assertNotNull(eventTrackerStatus);
+            OptionalLong optionalPosition = eventTrackerStatus.getCurrentPosition();
             assertTrue(optionalPosition.isPresent());
             assertEquals(totalNumberOfEvents, optionalPosition.getAsLong());
         });
+        processAnyDeadLettersPeriodically();
```

Comment:
It has something to do with the transaction and unique constraints. Moving it helped. The exception is:

```
Caused by: java.sql.SQLIntegrityConstraintViolationException: integrity constraint violation: unique constraint or index violation; SYS_PK_10102 table: DEADLETTERENTRY
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 34 more
```

I've tried many things. Starting the dead letter thread after the processor seemed to help. It might be a timing issue

Reply:
This is the full logs btw:

```
2022-09-07 10:53:12,627 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Storing DeadLetter (id: [b3a1ba4a-464e-45fb-b307-2c93f6d5f872]) for sequence [0] with index [0] in processing group [problematicProcessingGroup].
2022-09-07 10:53:12,628 [pool-4-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - No claimable and/or matching dead letters found to process.
2022-09-07 10:53:12,651 [pool-4-thread-1] ERROR SqlExceptionHelper             [] - integrity constraint violation: unique constraint or index violation; SYS_PK_10102 table: DEADLETTERENTRY
org.axonframework.messaging.IllegalPayloadAccessException: This result completed exceptionally, payload is not available. Try calling 'exceptionResult' to see the cause of failure.
	at org.axonframework.messaging.GenericResultMessage.getPayload(GenericResultMessage.java:189)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventHandlerInvoker.process(DeadLetteringEventHandlerInvoker.java:160)
	at org.axonframework.messaging.deadletter.SequencedDeadLetterProcessor.processAny(SequencedDeadLetterProcessor.java:40)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventIntegrationTest.processAnyDeadLetter(DeadLetteringEventIntegrationTest.java:181)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1626)
	at org.hibernate.query.Query.getResultList(Query.java:165)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.lambda$refreshPageIfNecessary$0(PagingJpaQueryIterable.java:102)
	at org.axonframework.common.transaction.TransactionManager.executeInTransaction(TransactionManager.java:47)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.refreshPageIfNecessary(PagingJpaQueryIterable.java:98)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.hasNext(PagingJpaQueryIterable.java:80)
	at org.axonframework.eventhandling.deadletter.jpa.JpaSequencedDeadLetterQueue.process(JpaSequencedDeadLetterQueue.java:305)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventHandlerInvoker.lambda$process$1(DeadLetteringEventHandlerInvoker.java:160)
	at org.axonframework.messaging.unitofwork.DefaultUnitOfWork.executeWithResult(DefaultUnitOfWork.java:77)
	at org.axonframework.messaging.unitofwork.UnitOfWork.executeWithResult(UnitOfWork.java:329)
	... 10 more
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:59)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3908)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:57)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1372)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1452)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1649)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617)
	... 19 more
Caused by: java.sql.SQLIntegrityConstraintViolationException: integrity constraint violation: unique constraint or index violation; SYS_PK_10102 table: DEADLETTERENTRY
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 34 more
Caused by: org.hsqldb.HsqlException: integrity constraint violation: unique constraint or index violation; SYS_PK_10102 table: DEADLETTERENTRY
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.Constraint.getException(Unknown Source)
	at org.hsqldb.index.IndexAVL.insert(Unknown Source)
	at org.hsqldb.persist.RowStoreAVL.indexRow(Unknown Source)
	at org.hsqldb.TransactionManager2PL.addInsertAction(Unknown Source)
	at org.hsqldb.Session.addInsertAction(Unknown Source)
	at org.hsqldb.Table.insertSingleRow(Unknown Source)
	at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
	at org.hsqldb.StatementInsert.getResult(Unknown Source)
	at org.hsqldb.StatementDMQL.execute(Unknown Source)
	at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
	at org.hsqldb.Session.execute(Unknown Source)
	... 37 more
2022-09-07 10:53:12,657 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Adding dead letter [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=true}}, metadata={}, messageIdentifier='b642e9cd-3bb4-4094-aa06-b9e04b7475b3', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=7}}}] because the sequence identifier [0] is already present.
2022-09-07 10:53:12,658 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Storing DeadLetter (id: [11399e26-56d5-41e4-a56f-ee95b7243e93]) for sequence [0] with index [1] in processing group [problematicProcessingGroup].
2022-09-07 10:53:12,658 [pool-3-thread-1] INFO  adLetteringEventHandlerInvoker [] - Event [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=true}}, metadata={}, messageIdentifier='b642e9cd-3bb4-4094-aa06-b9e04b7475b3', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=7}}}] is added to the dead-letter queue since its queue id [0] is already present.
2022-09-07 10:53:12,661 [pool-4-thread-1] ERROR SqlExceptionHelper             [] - statement is closed
org.axonframework.messaging.IllegalPayloadAccessException: This result completed exceptionally, payload is not available. Try calling 'exceptionResult' to see the cause of failure.
	at org.axonframework.messaging.GenericResultMessage.getPayload(GenericResultMessage.java:189)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventHandlerInvoker.process(DeadLetteringEventHandlerInvoker.java:160)
	at org.axonframework.messaging.deadletter.SequencedDeadLetterProcessor.processAny(SequencedDeadLetterProcessor.java:40)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventIntegrationTest.processAnyDeadLetter(DeadLetteringEventIntegrationTest.java:181)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not insert: [org.axonframework.eventhandling.deadletter.jpa.DeadLetterEntry]
2022-09-07 10:53:12,661 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Adding dead letter [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=true}}, metadata={}, messageIdentifier='92b6153d-c8e0-467b-ba8d-8bde1473c760', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=8}}}] because the sequence identifier [0] is already present.
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1626)
	at org.hibernate.query.Query.getResultList(Query.java:165)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.lambda$refreshPageIfNecessary$0(PagingJpaQueryIterable.java:102)
	at org.axonframework.common.transaction.TransactionManager.executeInTransaction(TransactionManager.java:47)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.refreshPageIfNecessary(PagingJpaQueryIterable.java:98)
	at org.axonframework.common.jpa.PagingJpaQueryIterable$PagingIterator.hasNext(PagingJpaQueryIterable.java:80)
	at org.axonframework.eventhandling.deadletter.jpa.JpaSequencedDeadLetterQueue.process(JpaSequencedDeadLetterQueue.java:305)
	at org.axonframework.eventhandling.deadletter.DeadLetteringEventHandlerInvoker.lambda$process$1(DeadLetteringEventHandlerInvoker.java:160)
	at org.axonframework.messaging.unitofwork.DefaultUnitOfWork.executeWithResult(DefaultUnitOfWork.java:77)
	at org.axonframework.messaging.unitofwork.UnitOfWork.executeWithResult(UnitOfWork.java:329)
	... 10 more
Caused by: org.hibernate.exception.SQLGrammarException: could not insert: [org.axonframework.eventhandling.deadletter.jpa.DeadLetterEntry]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:103)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3399)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3908)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:57)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1372)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1452)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1649)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617)
	... 19 more
Caused by: java.sql.SQLException: statement is closed
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCStatementBase.checkClosed(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.checkSetParameterIndex(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.setParameter(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.setBytes(Unknown Source)
	at org.hibernate.type.descriptor.sql.BlobTypeDescriptor$3$1.doBind(BlobTypeDescriptor.java:113)
	at org.hibernate.type.descriptor.sql.BlobTypeDescriptor$2$1.doBind(BlobTypeDescriptor.java:86)
	at org.hibernate.type.descriptor.sql.BasicBinder.bind(BasicBinder.java:73)
	at org.hibernate.type.AbstractStandardBasicType.nullSafeSet(AbstractStandardBasicType.java:276)
	at org.hibernate.type.AbstractStandardBasicType.nullSafeSet(AbstractStandardBasicType.java:271)
	at org.hibernate.type.AbstractSingleColumnStandardBasicType.nullSafeSet(AbstractSingleColumnStandardBasicType.java:39)
	at org.hibernate.persister.entity.AbstractEntityPersister.dehydrate(AbstractEntityPersister.java:3073)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3372)
	... 32 more
Caused by: org.hsqldb.HsqlException: statement is closed
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	... 45 more
2022-09-07 10:53:12,662 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Storing DeadLetter (id: [0038adea-2fb4-4bab-b69e-a9bed18d167d]) for sequence [0] with index [2] in processing group [problematicProcessingGroup].
2022-09-07 10:53:12,662 [pool-3-thread-1] INFO  adLetteringEventHandlerInvoker [] - Event [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=true}}, metadata={}, messageIdentifier='92b6153d-c8e0-467b-ba8d-8bde1473c760', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=8}}}] is added to the dead-letter queue since its queue id [0] is already present.
2022-09-07 10:53:12,665 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Adding dead letter [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=false}}, metadata={}, messageIdentifier='003b6032-78fa-4689-ae59-87f76b20dd72', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=9}}}] because the sequence identifier [0] is already present.
2022-09-07 10:53:12,665 [pool-3-thread-1] INFO  JpaSequencedDeadLetterQueue    [] - Storing DeadLetter (id: [ed63c73b-e649-4522-9b3a-0a34b392f72e]) for sequence [0] with index [3] in processing group [problematicProcessingGroup].
2022-09-07 10:53:12,665 [pool-3-thread-1] INFO  adLetteringEventHandlerInvoker [] - Event [GenericTrackedEventMessage{payload={DeadLetterableEvent{aggregateIdentifier='0', shouldSucceed=true, shouldSucceedOnEvaluation=false}}, metadata={}, messageIdentifier='003b6032-78fa-4689-ae59-87f76b20dd72', timestamp='2022-09-07T08:53:12.163Z, trackingToken={IndexTrackingToken{globalIndex=9}}}] is added to the dead-letter queue since its queue id [0] is already present.
2022-09-07 10:53:12,668 [pool-4-thread-1] ERROR SqlExceptionHelper             [] - statement is closed
org.axonframework.messaging.IllegalPayloadAccessException: This result completed exceptionally, payload is not available. Try calling 'exceptionResult' to see the cause of failure.
```

Reply:
Decided together to move this to the `axon-spring` module and use Spring transaction management. This works!